File: BOJ/[2800] 괄호 제거/KingDonggyu/2800.js
function solution(mathExp) {
  let answer = new Set();

  function solutionInner(mathExp) {
    if (answer.has(mathExp)) {
      return;
    }
    
    answer.add(mathExp);
    mathExp = Array.from(mathExp);
    
    const stack = [];

    mathExp.forEach((ch, i) => {
      if (ch === '(') {
        stack.push(i);
        return;
      }
      if (ch === ')') {
        temp = [...mathExp];
        temp[stack.pop()] = temp[i] = '';
        solutionInner(temp.join(''));
      }
    });
  }

  solutionInner(mathExp);
  answer = Array.from(answer).sort();
  return answer.slice(1).join('\n');
}

const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

rl.on('line', (line) => {
  console.log(solution(line));
  rl.close();
}).on('close', () => {
  process.exit();
});

File: BOJ/[2800] 괄호 제거/JFe/2800.py
import sys
from itertools import combinations

s = sys.stdin.readline()
stack, arr, answer = [], [], set()
# 괄호 인덱스 저장
for i, c in enumerate(s):
    if c == '(': stack.append(i)
    elif c == ')': arr.append([stack.pop(), i])
# 인덱스별 조합 생성 후 괄호 제거
for i in range(1, len(arr)+1):
    comb = list(combinations(arr, i))
    for c in comb:
        temp = list(s)
        for a, b in c:
            temp[a], temp[b] = '', ''
        answer.add(''.join(temp[:-1]))
for a in sorted(answer): print(a)